### PR TITLE
build(deps): replace hug with fastapi

### PR DESCRIPTION
**Pull request by afuetterer**
This PR proposes:

- replace hug with fastapi, the standard way to build REST APIs in Python these days
- add a pre-commit config, including ruff
- add a code coverage report to the github actions summary 

This will allow to use Python 3.12 as well.

### Tasks
- [x] add pydantic types (see: https://github.com/dracor-org/dracor-api/blob/630a7504c2c12750e2437006a7bd82a794cd4399/api.yaml#L1557)
- [x] fix tests
- [x] add coverage report in github actions summary

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -82,7 +84,3 @@ def metrics(segments):
         'numEdges': G.number_of_edges(),
         'nodes': nodes
     }
-
-
-if __name__ == '__main__':
-    metrics.interface.cli()
```
Comment (afuetterer) [inline app/main.py:88]: Did you use this feature of hug often? To use the API as a CLI?

File: README.md [hunk 2]
```diff
@@ -2,17 +2,13 @@
...
-supported Python version for now.*
-
 ## Getting started
 
 ```bash
 git clone https://github.com/dracor-org/dracor-metrics.git
 cd dracor-metrics
 poetry install
-poetry run hug -f app/main.py -p 8030
+poetry run uvicorn app.main:app --reload --port 8030
```
Comment (afuetterer) [inline README.md:11]: The starts a development server locally, that accepts requests on localhost:8030.

File: Dockerfile [hunk 3]
```diff
@@ -1,12 +1,10 @@
...
 
 WORKDIR /dracor-metrics
 
 COPY . .
-RUN pip install poetry && poetry install
+RUN pip install poetry && poetry install --without dev
 
 EXPOSE 8030
 
-CMD [ "poetry", "run", "hug", "-f", "app/main.py", "-p", "8030" ]
```
Comment (afuetterer) [inline Dockerfile:12]: This was running the development server in the Docker image, right?

File: Dockerfile [hunk 4]
```diff
@@ -1,12 +1,10 @@
...
+CMD [ "poetry", "run", "uvicorn", "app.main:app", "--host", "0.0.0.0", "--port", "8030" ]
```
Comment (afuetterer) [inline Dockerfile:10]: This is not ideal yet. But it works.

File: Dockerfile (hunk 3 above)
Comment (cmil) [reply Dockerfile:12]: This was running the service in the docker container.

File: app/main.py (hunk 1 above)
Comment (cmil) [reply app/main.py:88]: No, we did not use this at all.

File: Dockerfile (hunk 4 above)
Comment (cmil) [reply Dockerfile:10]: Not ideal because of `0.0.0.0`?

File: app/main.py (hunk 1 above)
Comment (afuetterer) [reply app/main.py:88]: Alright.

File: Dockerfile (hunk 3 above)
Comment (afuetterer) [reply Dockerfile:12]: Yes, it this runs a development server nevertheless? I am not familiar with hug. But my understanding was, that most web frameworks run a local development server when invoked.
The recommendation for production mostly includes using another WSGI/ASGI server like Gunicorn or uwsgi.

File: Dockerfile (hunk 4 above)
Comment (afuetterer) [reply Dockerfile:10]: Sorry, I should have elaborated more.
The FastAPI docs recommend using uvicorn in combination with gunicorn. But so far this is working.

https://fastapi.tiangolo.com/deployment/server-workers/

Comment (cmil) [reply Dockerfile:10]: We can probably do it as a follow up.

File: Dockerfile (hunk 3 above)
Comment (cmil) [reply Dockerfile:12]: Yes, I think it is the same principle. It's running a dev server that is not really meant for production use.

**Comment by afuetterer** [comment]
Ready for review @cmil.

FastAPI also includes its own auto-generated docs at: http://127.0.0.1:8030/docs

File: app/main.py [hunk 5]
```diff
@@ -59,30 +60,27 @@ def metrics(segments):
...
-        'maxDegreeIds': max_degree_ids,
-        'numConnectedComponents': nx.number_connected_components(G),
-        'numEdges': G.number_of_edges(),
-        'nodes': nodes
-    }
-
-
-if __name__ == '__main__':
-    metrics.interface.cli()
+            nodes[n].eigenvector = ec[n]
```
Comment (afuetterer) [inline app/main.py:72]: I was unsure about this conditional setting of the eigenvector value.

I used eigenvector=None above. What was the output of the hug based API here? Did the nodeinplaymetrics simply not have an eigenvector key?

Is it supposed to be None, zero or not having the key?

FastAPI and its underlying Pydantic are very strict regarding this. If you want to init a NodeInPlayMetrics without an eigenvector value, it fails.

File: app/main.py [hunk 6]
```diff
@@ -1,28 +1,31 @@
...
 
+@app.get("/")
+async def root() -> ServiceMetadata:
+    return ServiceMetadata(version=__version__)
 
-@hug.post('/metrics')
-def metrics(segments):
+
+@app.post("/metrics")
+async def metrics(segments: Segments) -> PlayMetrics:
```
Comment (afuetterer) [inline app/main.py:21]: This validates the incoming JSON.

**Comment by afuetterer** [comment]
What do you think?

**Comment by afuetterer** [comment]
Should I close this?

**Comment by cmil** [comment]
@afuetterer Please leave it open. I just haven't had time yet to finish the review.

File: app/main.py (hunk 5 above)
Comment (cmil) [reply app/main.py:72]: Yes, the current implementation just omits the `eigenvector` property if the calculation fails. You can see this when posting the response from the above mentioned URL to the metrics service (assuming its running on localhost:8030):

```sh
http https://dracor.org/api/v1/corpora/rus/plays/lermontov-strannyj-chelovek | http :8030/metrics
```

To set it to `None`/`null` instead should be ok. I will check this with dracor-api after merging.

**Pull request merged**